File: src/main/java/binarysearch/SearchRotatedSortedArray.java
package binarysearch;

public class SearchRotatedSortedArray {
    public int sol(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (nums[mid] == target) {
                return mid;
            }

            if (nums[left] <= nums[mid]) {
                if (nums[mid] < target || target < nums[left]) {
                    left = mid + 1;  // go right
                } else {
                    right = mid - 1;
                }
            } else {
                if (target < nums[mid] || nums[right] < target) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }

        }
        return -1;
    }

    public static void main(String[] args) {
    }
}

File: src/main/python/SearchRotatedSortedArray.py
from typing import List


class Solution:
    def search_iteratively(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        while left <= right:
            mid = left + int((right - left) / 2)
            if nums[mid] == target:
                return mid
            # To find whether pivot is located at the left or right side
            if nums[left] <= nums[mid]:  # Pivot point is at right side
                if nums[left] <= target < nums[mid]:
                    # go to left side
                    right = mid - 1
                else:
                    left = mid + 1
            else:  # Pivot point is at left side
                if nums[mid] < target <= nums[right]:
                    left = mid + 1
                else:
                    right = mid - 1
        return -1




